### PR TITLE
Add ecMulG2 precompile

### PR DESCRIPTION
**Pull request by ColoCarletti**
## Description

This pull request introduces a precompile for efficiently computing scalar multiplication of a point on the D-type sextic twist of the BN254 curve over Fp2. The curve is defined by the equation:

> $E' : y^2 = x^3 + (19485874751759354771024239261021720505790618469301721065564631296452457478373 + 266929791119991161246907387137283842545076965332900288569378510910307636690 u)$

The addition operation involves a point where each coordinate $(X, Y)$ is represented as elements of Fp2 and a sacalar $k$.

For example:

$A = ( X, Y)$ 

Where:

$X = (aX_0 + aX_1 * u)$ 
$Y = (aY_0 + aY_1 * u)$ 

The calldata expects input parameters:

$aX_0,aX_1,aY_0,aY_1,k$

And returns:

$cX_0,cX_1,cY_0,cY_1$

## Test

First run the local node

```
make run-node
```

Once the node us up run:

```
make test PRECOMPILE=ecmul_g2
```

### CODE REVIEW

File: precompiles/EcMulG2.yul [hunk 1]
```diff
@@ -0,0 +1,611 @@
+object "EcMulG2" {
```
Comment (fkrause98) [inline precompiles/EcMulG2.yul:1]: Let's also add the sage reference script for this one @ColoCarletti.

File: precompiles/EcMulG2.yul [hunk 2]
```diff
@@ -0,0 +1,611 @@
...
+				// will burn the provided gas by calling this function.
+				precompileCall(0, gas())
+		  	}
+
+            /// @notice Calculate the bit length of a number.
+            /// @param x The number to calculate the bit length of.
+            /// @return ret The bit length of the number.
+            function bitLen(x) -> ret {
+                ret := 0
+                for {} x {} {
```
Comment (fkrause98) [inline precompiles/EcMulG2.yul:70]: This for loop condition is a bit weird, this loops until x is zero?

**Comment by fkrause98** [comment]
Some tests are failing for me, I've double checked cloning from zero and trying to run them, 
maybe there are some files without commiting?

```
cd tests && \
cargo test ecmul_g2

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 36 filtered out; finished in 0.00s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 196 filtered out; finished in 0.00s


running 10 tests
test ecmul_g2_p_times_two_is_2p ... FAILED
test ecmul_g2_p_infinity_times_0_is_infinity ... FAILED
test ecmul_g2_p_infinity_times_k_is_infinity ... FAILED
test ecmul_g2_p_times_one_is_p ... FAILED
test ecmul_g2_valid_1 ... FAILED
test ecmul_g2_invalid_p_not_in_curve ... ok
test ecmul_g2_p_times_zero_is_infinity ... FAILED
test ecmul_g2_valid_2 ... FAILED
test ecmul_g2_valid_3 ... FAILED
test ecmul_g2_coordinates_not_in_field ... ok

failures:

---- ecmul_g2_p_times_two_is_2p stdout ----
thread 'ecmul_g2_p_times_two_is_2p' panicked at tests/test_utils.rs:40:10:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- ecmul_g2_p_infinity_times_0_is_infinity stdout ----
thread 'ecmul_g2_p_infinity_times_0_is_infinity' panicked at tests/test_utils.rs:40:10:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }

---- ecmul_g2_p_infinity_times_k_is_infinity stdout ----
thread 'ecmul_g2_p_infinity_times_k_is_infinity' panicked at tests/test_utils.rs:40:10:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }

---- ecmul_g2_p_times_one_is_p stdout ----
thread 'ecmul_g2_p_times_one_is_p' panicked at tests/test_utils.rs:40:10:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }

---- ecmul_g2_valid_1 stdout ----
thread 'ecmul_g2_valid_1' panicked at tests/test_utils.rs:40:10:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }

---- ecmul_g2_p_times_zero_is_infinity stdout ----
thread 'ecmul_g2_p_times_zero_is_infinity' panicked at tests/test_utils.rs:40:10:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }

---- ecmul_g2_valid_2 stdout ----
thread 'ecmul_g2_valid_2' panicked at tests/test_utils.rs:40:10:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }

---- ecmul_g2_valid_3 stdout ----
thread 'ecmul_g2_valid_3' panicked at tests/test_utils.rs:40:10:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }


failures:
    ecmul_g2_p_infinity_times_0_is_infinity
    ecmul_g2_p_infinity_times_k_is_infinity
    ecmul_g2_p_times_one_is_p
    ecmul_g2_p_times_two_is_2p
    ecmul_g2_p_times_zero_is_infinity
    ecmul_g2_valid_1
    ecmul_g2_valid_2
    ecmul_g2_valid_3

test result: FAILED. 2 passed; 8 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.12s


```

**Review by fkrause98 — APPROVED** [review]
LGTM!

**Comment by IAvecilla** [comment]
Can we change this `unwrap` in the `test_utils` file:
```rust
fn write_line_to_report(used_gas: u32, report_to_write: &str) {
    let mut file = OpenOptions::new()
        .append(true)
        .open(report_to_write)
        .unwrap()
        ...
}
```
to something like:

```rust
fn write_line_to_report(used_gas: u32, report_to_write: &str) {
    let mut file = OpenOptions::new()
        .append(true)
        .open(report_to_write)
        .expect(&format!("Unable to write the gas report in {report_to_write:?}"));
        ...
}
```

Just to better indicate that the error happens in the gas reports generation and make it easier to understand.

**Review by IAvecilla — APPROVED** [review]

**Pull request merged**